agent/pkg/singularity/singularity.go

```diff
@@ -0,0 +1,337 @@
+	for _, env := range req.ContainerConfig.Env {
+		args = append(args, "--env", env)
+	}
+
+	tmpdir, err := os.MkdirTemp("/var/tmp", fmt.Sprintf("*-%s", id)) // TODO: cleanup
+	if err != nil {
+		return "", fmt.Errorf("making tmp dir for archives: %w", err)
+	}
+
+	req.Archives = append(req.Archives)
+
+	for _, a := range req.Archives {
+		src := filepath.Join(tmpdir, a.Path)
+		if err := archive.Write(src, a.Archive); err != nil {
+			return "", fmt.Errorf("writing archive for %s: %w", a.Path, err)
+		}
+	}
+	args = append(args, "--bind", fmt.Sprintf("%s:%s", tmpdir, containerArchivePath))
+
+	// TODO: device mappings and stuff for amd.
+	for _, d := range req.HostConfig.DeviceRequests {
+		if d.Driver == "nvidia" {
+			args = append(args, "--nv")
+			break
+		}
+	}
+
+	args = append(args, req.ContainerConfig.Image)
+	args = append(args, path.Join(containerArchivePath, runDir, singularityWrapperEntrypoint))
+	args = append(args, req.ContainerConfig.Cmd...)
+
+	s.log.Info("singularity \\")
+	toPrint := ""
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "--") {
+			toPrint += " \\\n"
+			toPrint += "\t"
+			toPrint += arg
+		} else {
+			toPrint += " "
+			toPrint += arg
+		}
+	}
+	s.log.Info(toPrint)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.containers[id] = &SingularityContainer{
+		Cmd:         append([]string{"singularity"}, args...),
+		Req:         req,
+		NetworkMode: "host",
+		Ports:       req.ContainerConfig.ExposedPorts,
+		TmpDir:      tmpdir,
+	}
+	return id.String(), nil
+}
+
+// RunContainer implements container.ContainerRuntime
+func (s *SingularityClient) RunContainer(
+	ctx context.Context,
+	waitCtx context.Context,
+	id string,
+	p events.Publisher[docker.Event],
+) (*docker.Container, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var cont *SingularityContainer
+	for cID, rcont := range s.containers {
+		if cproto.ID(id) != cID {
+			continue
+		}
+		cont = rcont
+		break
+	}
+	if cont == nil {
+		return nil, container.ErrMissing
+	}
+
+	cmd := exec.CommandContext(waitCtx, cont.Cmd[0], cont.Cmd[1:]...) // TODO: no command ctx, recover stuff
+	stdout, oerr := cmd.StdoutPipe()
+	stderr, eerr := cmd.StderrPipe()
+	if oerr != nil || eerr != nil {
+		s.log.Error(oerr.Error(), eerr.Error())
+	} else {
+		s.wg.Go(func(ctx context.Context) {
+			for scan := bufio.NewScanner(stdout); scan.Scan(); {
+				p.Publish(ctx, docker.NewLogEvent(model.LogLevelInfo, scan.Text())) // TODO: stdtype
+			}
+		})
+		s.wg.Go(func(ctx context.Context) {
+			for scan := bufio.NewScanner(stderr); scan.Scan(); {
+				p.Publish(ctx, docker.NewLogEvent(model.LogLevelInfo, scan.Text())) // TODO: stdtype
+			}
+		})
+	}
+
+	// TODO: device mappings and stuff for amd.
+	var devices string
+	for _, d := range cont.Req.HostConfig.DeviceRequests {
+		if d.Driver == "nvidia" {
+			devices = strings.Join(d.DeviceIDs, ",")
+		}
+	}
+	cmd.Env = append(cmd.Env,
+		fmt.Sprintf("SINGULARITYENV_CUDA_VISIBLE_DEVICES=%s", devices),
```

Comment:
The agent can run multiple containers at a time?

And devices accommodates that by allocating individual GPUs to a container, I presume?

Reply:
yes exactly